splay/src/main/scala/org/scalajs/benchmark/splay/Splay.scala

```diff
@@ -0,0 +1,447 @@
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Google Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// This benchmark is based on a JavaScript log processing module used
+// by the V8 profiler to generate execution time profiles for runs of
+// JavaScript applications, and it effectively measures how fast the
+// JavaScript engine is at allocating nodes and reclaiming the memory
+// used for old nodes. Because of the way splay trees work, the engine
+// also has to deal with a lot of changes to the large tree object
+// graph.
+
+// This is a Scala implementation of the Splay benchmark from:
+// https://github.com/chromium/octane/blob/master/splay.js
+
+// Ported by Nicolas Stucki to Scala.js.
+
+package org.scalajs.benchmark.splay
+
+import scala.annotation.tailrec
+import scala.scalajs.js
+
+object Splay extends org.scalajs.benchmark.Benchmark {
+  override def prefix: String = "Splay"
+
+  def run(): Unit = {
+    resetRandomSeed()
+    splaySetup()
+    splayRun()
+    splayTearDown()
+  }
+
+  //// Configuration.
+  private final val kSplayTreeSize = 8000
+  private final val kSplayTreeModifications = 80
+  private final val kSplayTreePayloadDepth = 5
+
+  private var splayTree: SplayTree[js.Dynamic] = null
+  private var splaySampleTimeStart = 0L
+
+  private def generatePayloadTree(depth: Int, tag: String): js.Dynamic = {
+    if (depth == 0) {
+      js.Dynamic.literal(
+        array = js.Array(0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
+        string = "String for key " + tag + " in leaf node"
```

Comment:
The bottleneck was here. Before I had `string = "String for key $tag in leaf node"` and the benchmarks gave:
```
splay [all] sbt           :        14.86 real        54.88 user         2.04 sys
splay [fastopt] d8        : Splay: 189363.63636363635 us
splay [fastopt] node      : Splay: 170916.66666666666 us
splay [fullopt] d8        : Splay: 218800 us
splay [fullopt] node      : Splay: 150000 us
splay [js] d8             : Splay(RunTime): 46045.454545454544 us
splay [js] node           : Splay(RunTime): 47295.454545454544 us
```
against the current:
```
splay [all] sbt           :        11.54 real        43.59 user         1.84 sys
splay [fastopt] d8        : Splay: 45590.90909090909 us
splay [fastopt] node      : Splay: 45909.09090909091 us
splay [fullopt] d8        : Splay: 43702.12765957447 us
splay [fullopt] node      : Splay: 45177.77777777778 us
splay [js] d8             : Splay(RunTime): 43717.391304347824 us
splay [js] node           : Splay(RunTime): 50333.333333333336 us
```

So 3~4X slower because of the string interpolator.